src/H5FDint.c

```diff
@@ -260,152 +290,2049 @@ H5FD_write(H5FD_t *file, H5FD_mem_t type, haddr_t addr, size_t size, const void
+ *              vectors in *s_types_ptr, *s_addrs_ptr, *s_sizes_ptr, and
+ *              *s_bufs_ptr respectively.  Determine the sorted order
+ *              of the vector I/O request, and load it into the new
+ *              vectors in sorted order.
+ *
+ *              Note that in this case, it is the callers responsibility
+ *              to free the sorted vectors.
+ *
+ *                                            JRM -- 3/15/21
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+
+static int
+H5FD__vsrt_tmp_cmp(const void *element_1, const void *element_2)
+{
+    haddr_t addr_1    = ((const H5FD_vsrt_tmp_t *)element_1)->addr;
+    haddr_t addr_2    = ((const H5FD_vsrt_tmp_t *)element_2)->addr;
+    int     ret_value = 0; /* Return value */
+
+    FUNC_ENTER_STATIC_NOERR
+
+    /* Sanity checks */
+    HDassert(H5F_addr_defined(addr_1));
+    HDassert(H5F_addr_defined(addr_2));
+
+    if (H5F_addr_gt(addr_1, addr_2)) {
+
+        ret_value = 1;
+    }
+    else if (H5F_addr_lt(addr_1, addr_2)) {
+
+        ret_value = -1;
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* H5FD__vsrt_tmp_cmp() */
+
+herr_t
+H5FD_sort_vector_io_req(hbool_t *vector_was_sorted, uint32_t count, H5FD_mem_t types[], haddr_t addrs[],
+                        size_t sizes[], const void *bufs[], H5FD_mem_t **s_types_ptr, haddr_t **s_addrs_ptr,
+                        size_t **s_sizes_ptr, void ***s_bufs_ptr)
+{
+    herr_t                  ret_value = SUCCEED; /* Return value */
+    int                     i;
+    struct H5FD_vsrt_tmp_t *srt_tmp = NULL;
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    /* Sanity checks */
+
+    HDassert(vector_was_sorted);
+
+    HDassert((types) || (count == 0));
+    HDassert((addrs) || (count == 0));
+    HDassert((sizes) || (count == 0));
+    HDassert((bufs) || (count == 0));
+
+    /* verify that the first elements of the sizes and types arrays are
+     * valid.
+     */
+    HDassert((count == 0) || (sizes[0] != 0));
+    HDassert((count == 0) || (types[0] != H5FD_MEM_NOLIST));
+
+    HDassert((count == 0) || ((s_types_ptr) && (NULL == *s_types_ptr)));
+    HDassert((count == 0) || ((s_addrs_ptr) && (NULL == *s_addrs_ptr)));
+    HDassert((count == 0) || ((s_sizes_ptr) && (NULL == *s_sizes_ptr)));
+    HDassert((count == 0) || ((s_bufs_ptr) && (NULL == *s_bufs_ptr)));
+
+    *vector_was_sorted = TRUE;
+
+    /* if count <= 1, vector is sorted by definition */
+    if (count > 1) {
+
+        /* scan the addrs array to see if it is sorted */
+        i = 1;
+
+        while ((*vector_was_sorted) && (i < (int)(count - 1))) {
+
+            if (H5F_addr_gt(addrs[i - 1], addrs[i])) {
+
+                *vector_was_sorted = FALSE;
+            }
+            else if (H5F_addr_eq(addrs[i - 1], addrs[i])) {
+
+                HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "duplicate addr in vector")
+            }
+            i++;
+        }
+    }
```

Comment:
The iteration pattern is more clear if you write it in "headline" form using a for-loop. Starting right before `*vector_was_sorted = TRUE;` you can write,

```C
        assert(count < 1 || H5F_addr_defined(addrs[0]));

        /* See if the addresses are in ascending order and unique. */
        for (i = 1; i < count; i++) {
                assert(H5F_addr_defined(addrs[i]));
                if (addrs[i - 1] > addrs[i])
                        break;
                if (addrs[i - 1] == addrs[i])
                        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "duplicate addr in vector")
        }

        if (count <= 1 || i == count)
                *vector_was_sorted = TRUE;
        else
                *vector_was_sorted = FALSE;
```

No further defined-ness checking (e.g., in qsort) should be necessary.

Reply:
Incidentally, isn't there a bug in the while-loop?

Reply:
I can see the for loop, though I think assigning vector_was_sorted after the loop is unnecessarily opaque, just assign it in the loop then break. Why do you want to get rid of H5F_addr_gt/eq? Is the bug that it should be (i < count) not (i < count - 1)?

Reply:
Again, @fortnern is correctly using the address comparison macros here.   haddr_t are designed to be extensible to 128-bits (soon?)  and the address comparison macros help that path.